### PR TITLE
RandomDataGenerator Mapping, Delisting, Splits, and Dividends Generator

### PR DESCRIPTION
**Pull request by gsalaz98**
Create mapping, IPO, delisting, splits, and dividends random generation

#### Description

Created random-seed argument for rdg in order to let the user get deterministic output

Update documentation in 'FactorFileRow.cs' to accurately reflect factor
file structure

Update CSV generation for FactorFile so that it uses FactorFileRow's CSV
generator

Add FinancialCalendar to make it easier to implement logic regarding
financial quarters

Add mapping events to RandomDataGenerator

Update MapFileRow ToCsv method to correctly emit the ticker as lowercase

Fixed bug in FactorFile where we would get the same initial data point twice
when we converted it to CSV

Create new method to convert a MapFile to CSV

Create new method to write MapFile to disk as CSV

Add unit test to test for successful CSV generation in MapFile

Add new files to project

Add FinancialCalendar unit tests

Create new class to handle generation of dividends, splits, and maps

#### Related Issue

#2859 
#### Motivation and Context

Allows the user to generate data for backtest benchmarking and testing

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

Add new unit tests to cover some changes

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by Jay-Jay-D** [comment]
Seems pretty solid PR Gerardo.

However, I cannot test its behavior, the CLI is not working as expected.

image

Also, I'd suggest to explicit the default values for the optional arguments.

File: Common/Data/Auxiliary/MapFile.cs [hunk 1]
```diff
@@ -131,6 +143,16 @@ public static MapFile Read(string permtick, string market)
             return new MapFile(permtick, MapFileRow.Read(permtick, market));
         }
 
+        /// <summary>
+        /// Writes the map file to a CSV file
+        /// </summary>
+        /// <param name="market">The market to save the MapFile to</param>
+        public void WriteToCsv(string market) 
+        {
+            var filePath = GetMapFilePath(Permtick, market);
```
Comment (Jay-Jay-D) [inline Common/Data/Auxiliary/MapFile.cs]: Maybe we can heck if the folder exist first.

File: ToolBox/RandomDataGenerator/DividendSplitMapGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        /// </summary>
+        /// <param name="tickHistory"></param>
+        public void GenerateSplitsDividends(IEnumerable<Tick> tickHistory)
+        {
+            var previousMonth = -1;
+            var monthsTrading = 0;
+
+            var hasRename = _randomValueGenerator.NextBool(30.0);
+            var hasSplits = _randomValueGenerator.NextBool(15.0);
+            var hasDividends = _randomValueGenerator.NextBool(60.0);
```
Comment (Jay-Jay-D) [inline ToolBox/RandomDataGenerator/DividendSplitMapGenerator.cs]: Is there any reason for those values? 

Maybe they can be parameterized?

**Comment by Jay-Jay-D** [comment]
Great Job @gsalaz98!

Just a couple of details.

File: Common/Data/Auxiliary/MapFile.cs (hunk 1 above)
Comment (gsalaz98) [reply Common/Data/Auxiliary/MapFile.cs]: What kind of behavior should we be aiming for here? I'm thinking about writing an error message to the log and returning with no write attempted if the directory can't be found

Comment (Jay-Jay-D) [reply Common/Data/Auxiliary/MapFile.cs]: I think that with a simple `Directory.Create(path)` is enough. If there is no folder it'll be created, if it exists, nothing happens

**Pull request merged**